---
title: "Functional differences between wild mice microbiome and SPF in ASO mice"
subtitle: "Data: HUMAnN 3.9 unstratified UniRef90 Gene Families"
author: "Daniel Dumitrescu, Giacomo Antonello, Nicola Segata, Levi Waldron, Sarkis Mazmanian"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
params:
  basic_outdir: "results"
  microbiome_basic_transform: "RelAbund"
  humann_data_type: "GeneFamilies"
  beta_dist: "bray"
  beta_MDS: "MDS"
---

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{r setup, include=FALSE}
# packages for tabular data manipulation
library(tidyverse)
library(data.table)

# microbiome specific packages
library(maaslin3)
library(mia)
# some packages for data visualization (plots and tables)
library(miaViz)
library(pheatmap)
library(UpSetR)
library(ggpubr)
library(kableExtra)
library(reactable)

# MAC's package to download data
library(parkinsonsMetagenomicData) # devtools::install_github("ASAP-MAC/parkinsonsMetagenomicData")
library(biobakeryUtils) # devtools::install_github("g-antonello/biobakeryUtils")
library(ggh4x) # for complex ggplot2 faceting
# set ggplot2 plot themes 
theme_set(theme_light())
# set seed for reproducibility
set.seed(1234)
```

# Create output directories all in one chunk

```{r}

output_directories.list <- list(
  # alpha diversity directory
  alpha_div_outdir = file.path(params$basic_outdir, params$humann_data_type, "01_alpha_diversity"),
  # beta diversity output directory
  beta_div_outdir = file.path(params$basic_outdir, params$humann_data_type, "02_beta_diversity")
)

tmp_out <- capture.output(sapply(output_directories.list, dir.create, showWarnings = FALSE, recursive = TRUE))
rm(tmp_out)
```

# Prepare data for analysis

```{r}
HUMAnN_genefamilies_unstratified.tse <- read_TSE_from_dir("Data/HUMAnN_genefamilies_unstratified.tse/")
HUMAnN_genefamilies_unstratified.tse
```

# Richness of gene families per microbiome type

```{r, fig.cap = "Histrogram of alpha indices with test for normality under each plot"}

alphaIndices <- c("dbp_dominance", "observed_richness", "shannon_diversity", "gini")

HUMAnN_genefamilies_unstratified.tse <- addAlpha(HUMAnN_genefamilies_unstratified.tse, assay.type = "gene_family_abundance_rpk_unstratified", index =  alphaIndices)

lapply(alphaIndices, 
       function(div) {
         stats_basic <- shapiro.test(as.data.frame(colData(HUMAnN_genefamilies_unstratified.tse))[[div]])
         text <- paste0("Shapiro-Wilk test: W = ", round(stats_basic$statistic, 2), "; P-value = ", round(stats_basic$p.value, 3))
         plot <- ggplot(as.data.frame(colData(HUMAnN_genefamilies_unstratified.tse)), aes(x = !!sym(div))) +
           geom_histogram() + 
           labs(
             title = div, 
             caption = text
           )
         
         }
       ) %>% 
  ggarrange(plotlist = .)


```

```{r, tab.cap="Assessment of dependence of alpha diversity measures in relation to sequencing depth"}
lm_alpha.list <- sapply(alphaIndices, function(div) {
  lm(as.formula(paste(div, "~", "donor_microbiome_type + sex + number_reads")), data = as.data.frame(colData(HUMAnN_genefamilies_unstratified.tse)))
}, USE.NAMES = TRUE, simplify = FALSE)

table_results.df <- lm_alpha.list %>% 
  lapply(broom::tidy) %>% 
  bind_rows(.id = "AlphaIndex") %>% 
  filter(!grepl("Intercept", term))

table_results.df %>%
  kbl() %>% 
  row_spec(which(table_results.df$p.value < 0.05), bold = TRUE) %>% 
  kable_styling()
```

# Compositional PCoA

```{r}
dimRedName <- paste(params$beta_dist, "mds", sep = "_")
# this will work once mia or TreeSummarizedExperiment fixes some things
HUMAnN_genefamilies_unstratified.tse <- mia::addMDS(
  HUMAnN_genefamilies_unstratified.tse,
  assay.type = "gene_family_abundance_rpk_unstratified",
  method = params$beta_dist,
  ncomponents = 3, 
  name = dimRedName
  )

variances_explained <- attr(reducedDim(HUMAnN_genefamilies_unstratified.tse, dimRedName), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

project1_meta_with_MDS.df <- cbind.data.frame(as.data.frame(colData(HUMAnN_genefamilies_unstratified.tse)), reducedDim(HUMAnN_genefamilies_unstratified.tse, dimRedName) %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

beta_plot_basic <- project1_meta_with_MDS.df %>% 
  ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = paste(sex, donor_microbiome_type))) + 
  geom_point(size = 2) + 
  stat_ellipse(show.legend = FALSE) + 
  #scale_color_manual(values = ggsci::pal_jco()(6)) +
  theme(legend.position = "top") +
  labs(color = "Sex and Microbiome",
       x = paste0("MDS 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("MDS 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

beta_plot_basic
```

```{r}
HUMAnN_genefamilies_unstratified.tse <- addPERMANOVA(
  HUMAnN_genefamilies_unstratified.tse,
  assay.type = "gene_family_abundance_rpk_unstratified",
  method = params$beta_dist,
  formula = x ~ sex + donor_microbiome_type,
  name = "permanova",
  permutations = 999,
  test.homogeneity = TRUE
)

permanova_table.df <- PERMANOVA_to_table(HUMAnN_genefamilies_unstratified.tse)

write_tsv(permanova_table.df, file = file.path(output_directories.list$beta_div_outdir, paste(params$beta_dist, "adonisPermanova", "sex", "donor_microbiome_type.tsv", sep = "_")))
```

```{r}
permanova_table.df %>% 
  kbl() %>% 
  kable_styling()
```


## Differential abundant gene families per microbiome type

### Sex adjusted

```{r}
# prepare input data

tmp.tse <- HUMAnN_genefamilies_unstratified.tse
tmp.tse <- tmp.tse[rowSums(assay(tmp.tse)) > 0 , ]
tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "WildR_vs_SPF_adj_sex"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$humann_data_type, maaslin3_expmt)
# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
WildR_vs_SPF_adj_sex.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ sex + donor_microbiome_type, 
    transform = "LOG", 
    normalization = "TSS",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE,
    # filter to have at least 100 counts in 2 samples
    min_abundance = 100, 
    min_prevalence = 2/ncol(tmp.tse),
    plot_associations = FALSE,
    save_plots_rds = TRUE,
    min_variance = 0.5, 
    cores = 8
  )

WildR_vs_SPF_adj_sex_curated.maaslin3 <- list(
  prevalence = WildR_vs_SPF_adj_sex.maaslin3$fit_prevalence$results,
  abundance = WildR_vs_SPF_adj_sex.maaslin3$fit_abundance$results
)

# write tables
write_maaslin3_curated_tables(WildR_vs_SPF_adj_sex_curated.maaslin3, out.dir = maaslin3_final_outdir)
# save raw output
saveRDS(WildR_vs_SPF_adj_sex.maaslin3, file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
# copy their .RDS ggplot object 
system(paste("cp", file.path(maaslin3_tmp_outdir, "figures", "summary_plot_gg.RDS"), maaslin3_final_outdir))

```

### Males

```{r}
# prepare input data

tmp.tse <- HUMAnN_genefamilies_unstratified.tse[colData(HUMAnN_genefamilies_unstratified.tse)$sex == "Male",]
tmp.tse <- tmp.tse[rowSums(assay(tmp.tse)) > 0 , ]
tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "WildR_vs_SPF_Males"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$humann_data_type, maaslin3_expmt)
# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
WildR_vs_SPF_Males.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ sex + donor_microbiome_type, 
    transform = "LOG", 
    normalization = "TSS",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE,
    # filter to have at least 100 counts in 2 samples
    min_abundance = 100, 
    min_prevalence = 2/ncol(tmp.tse),
    plot_associations = FALSE,
    save_plots_rds = TRUE,
    verbosity = "ERROR"
  )

WildR_vs_SPF_Males_curated.maaslin3 <- list(
  prevalence = WildR_vs_SPF_Males.maaslin3$fit_prevalence$results,
  abundance = WildR_vs_SPF_Males.maaslin3$fit_abundance$results
)

# write tables
write_maaslin3_curated_tables(WildR_vs_SPF_Males_curated.maaslin3, out.dir = maaslin3_final_outdir)
# save raw output
saveRDS(WildR_vs_SPF_Males.maaslin3, file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
# copy their .RDS ggplot object 
system(paste("cp", file.path(maaslin3_tmp_outdir, "figures", "summary_plot_gg.RDS"), maaslin3_final_outdir))

```

### Females

```{r}
# prepare input data

tmp.tse <- HUMAnN_genefamilies_unstratified.tse[colData(HUMAnN_genefamilies_unstratified.tse)$sex == "Female",]
tmp.tse <- tmp.tse[rowSums(assay(tmp.tse)) > 0 , ]
tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "WildR_vs_SPF_Females"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$humann_data_type, maaslin3_expmt)
# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
WildR_vs_SPF_Females.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ sex + donor_microbiome_type, 
    transform = "LOG", 
    normalization = "TSS",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE,
    # filter to have at least 100 counts in 2 samples
    min_abundance = 100, 
    min_prevalence = 2/ncol(tmp.tse),
    plot_associations = FALSE,
    save_plots_rds = TRUE,
    verbosity = "ERROR"
  )

WildR_vs_SPF_Females_curated.maaslin3 <- list(
  prevalence = WildR_vs_SPF_Females.maaslin3$fit_prevalence$results,
  abundance = WildR_vs_SPF_Females.maaslin3$fit_abundance$results
)

# write tables
write_maaslin3_curated_tables(WildR_vs_SPF_Females_curated.maaslin3, out.dir = maaslin3_final_outdir)
# save raw output
saveRDS(WildR_vs_SPF_Females.maaslin3, file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
# copy their .RDS ggplot object 
system(paste("cp", file.path(maaslin3_tmp_outdir, "figures", "summary_plot_gg.RDS"), maaslin3_final_outdir))

```

# Example information retrieval for a UniRef90 entry
use [uniPark entry mapper](https://search.r-project.org/CRAN/refmans/rbioapi/html/rba_uniprot_uniparc.html)
```{r}

```


# Session Info

```{r}
sessionInfo()
```

