---
title: "Functional differences between wild mice microbiome and SPF in ASO mice"
subtitle: "Data: HUMAnN 3.9 unstratified pathways"
author: "Daniel Dumitrescu, Giacomo Antonello, Nicola Segata, Levi Waldron, Sarkis Mazmanian"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
params:
  basic_outdir: "results"
  microbiome_basic_transform: "RelAbund"
  humann_data_type: "GeneFamilies"
  beta_dist: "bray"
  beta_MDS: "MDS"
---

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{r, include=FALSE}
# packages for tabular data manipulation
library(tidyverse)
library(data.table)

# microbiome specific packages
library(maaslin3)
library(mia)
# some packages for data visualization (plots and tables)
library(miaViz)
library(pheatmap)
library(UpSetR)
library(ggpubr)
library(kableExtra)
library(reactable)

# MAC's package to download data
library(parkinsonsMetagenomicData) # devtools::install_github("ASAP-MAC/parkinsonsMetagenomicData")
library(biobakeryUtils) # devtools::install_github("g-antonello/biobakeryUtils")
library(ggh4x) # for complex ggplot2 faceting
# set ggplot2 plot themes 
theme_set(theme_light())
# set seed for reproducibility
set.seed(1234)
```

# Create output directories all in one chunk

```{r}

output_directories.list <- list(
  # alpha diversity directory
  alpha_div_outdir = file.path(params$basic_outdir, params$humann_data_type, "01_alpha_diversity"),
  # beta diversity output directory
  beta_div_outdir = file.path(params$basic_outdir, params$humann_data_type, "02_beta_diversity"),
  # differential abundance temporary directory
  maaslin_tmpdir = "~/Documents/maaslin3_tmp_results/mazmanian_Danny"
)

tmp_out <- capture.output(sapply(output_directories.list, dir.create, showWarnings = FALSE, recursive = TRUE))
rm(tmp_out)
```

# Prepare data for analysis

THis is run only if interactive to avoid careless overwriting

```{r, eval=interactive(), message=FALSE}

# Load metadata
metadata.df <- read_tsv("Data/metadata_4.0.tsv") %>% 
  # format a few values in the metadata
  mutate(
    sex = as.factor(sex),
    cage = as.factor(cage),
    donor_microbiome_type = factor(donor_microbiome_type, levels = c("SPF", "WildR")),
    ) %>% 
  as.data.frame()

rownames(metadata.df) <- metadata.df$uuid

googleCloudStorageR::gcs_auth(json_file = "~/Downloads/curatedmetagenomicdata-232f4a306d1d.json")

# HUMAnN_3.9 pathway abundance unstratified
humann_unstrat_cache <- cacheMetagenomicData(metadata.df$uuid, data_type = "genefamilies_unstratified", redownload = "no")

# mpa_cache <- cacheMetagenomicData(metadata.df$uuid, data_type = "relative_abundance", redownload = "no") %>% 
#   # retrieve reads per sample too
#   mutate(n_reads_processed = get_reads_processed(cache_path))

humann_unstrat.se <- loadMetagenomicData(humann_unstrat_cache)
genefamilies_raw <- rownames(humann_unstrat.se)
# make initial rowData
rowData.df <- DataFrame(
  UiRef090_code = sapply(strsplit(genefamilies_raw, "_"), "[", 2),
  full_name = genefamilies_raw, 
  row.names = genefamilies_raw)

# fix NA in longer_name because of UNMAPPED/UNINTEGRATED
rowData.df$UiRef090_code[is.na(rowData.df$UiRef090_code)] <- rowData.df$full_name[is.na(rowData.df$UiRef090_code)]
# assign rowData
rowData(humann_unstrat.se) <- rowData.df
# identical(metadata.df$uuid, colnames(humann_unstrat.se)) # TRUE


# Create a de novo TreeSummarizedExperiment
InputHumannData_geneFamilies.tse <- TreeSummarizedExperiment(
  list("AbsAbund" = assay(humann_unstrat.se), 
       "AbsAbund_round" = round(assay(humann_unstrat.se), 0),
       "RelAbund" = apply(assay(humann_unstrat.se), 2, function(x) x/sum(x))
       ),
  rowData = rowData.df, 
  colData =  DataFrame(metadata.df))

# save this TSE object
dir.create("Data/humann_data", recursive = TRUE, showWarnings = FALSE)
write_tsv(as.data.frame(assay(InputHumannData_geneFamilies.tse, "AbsAbund")) %>% rownames_to_column("UniRef90_GeneFamily"), "Data/humann_data/humann3.9_unstratified_GeneFamilies_AbsCounts_raw.tsv")



saveRDS(GeneFamilies, "Data/InputHumannData_geneFamilies.tse.Rds")

```

```{r}
InputHumannData_geneFamilies.tse <- readRDS("Data/InputHumannData_geneFamilies.tse.Rds")
InputHumannData_geneFamilies.tse
```
